packages/wmr/src/lib/transform-prefresh-registrations.js

```diff
@@ -0,0 +1,452 @@
+					// This can happen if another Babel plugin replaces parents.
+					if (seenForRegistration.has(node)) {
+						return;
+					}
+					seenForRegistration.add(node);
+					// Don't mutate the tree above this point.
+
+					// export function Named() {}
+					// function Named() {}
+					findInnerComponents(inferredName, path, (persistentID, targetExpr) => {
+						const handle = createRegistration(programPath, persistentID);
+						insertAfterPath.insertAfter(t.expressionStatement(t.assignmentExpression('=', handle, targetExpr)));
+					});
+				}
+			},
+			VariableDeclaration(path) {
+				const node = path.node;
+				let programPath;
+				let insertAfterPath;
+
+				switch (path.parent.type) {
+					case 'Program':
+						insertAfterPath = path;
+						programPath = path.parentPath;
+						break;
+					case 'ExportNamedDeclaration':
+						insertAfterPath = path.parentPath;
+						programPath = insertAfterPath.parentPath;
+						break;
+					case 'ExportDefaultDeclaration':
+						insertAfterPath = path.parentPath;
+						programPath = insertAfterPath.parentPath;
+						break;
+					default:
+						return;
+				}
+
+				// Make sure we're not mutating the same tree twice.
+				// This can happen if another Babel plugin replaces parents.
+				if (seenForRegistration.has(node)) {
+					return;
+				}
+				seenForRegistration.add(node);
+				// Don't mutate the tree above this point.
+
+				const declPaths = path.get('declarations');
+				if (declPaths.length !== 1) {
+					return;
+				}
+
+				const declPath = declPaths[0];
+				const inferredName = declPath.node.id.name;
+				findInnerComponents(inferredName, declPath, (persistentID, targetExpr, targetPath) => {
+					if (targetPath === null) {
+						return;
+					}
+
+					const handle = createRegistration(programPath, persistentID);
+					if (targetPath.parent.type === 'VariableDeclarator') {
+						insertAfterPath.insertAfter(t.expressionStatement(t.assignmentExpression('=', handle, declPath.node.id)));
+					} else {
+						targetPath.replaceWith(t.assignmentExpression('=', handle, targetExpr));
+					}
+				});
+			},
+			Program: {
+				enter(path, state) {
+					state.set('filehash', hash(path.ctx.filename || 'unnamed'));
+					state.set('contexts', new Map());
+				},
+				exit(path) {
+					const registrations = registrationsByProgramPath.get(path);
+
+					if (registrations === undefined) {
+						return;
+					}
+
+					// Make sure we're not mutating the same tree twice.
+					// This can happen if another Babel plugin replaces parents.
+					const node = path.node;
+					if (seenForOutro.has(node)) {
+						return;
+					}
+					seenForOutro.add(node);
+					// Don't mutate the tree above this point.
+
+					registrationsByProgramPath.delete(path);
+					const declarators = [];
+					// TODO: currently this crashes somewhere due to a variableDeclaration missing an array of declarations
+					path.pushContainer('body', t.variableDeclaration('var', declarators));
```

Comment:
An issue presents itself here where we create a `VariableDeclaration` but the `declarations` array doesn't get populated.

This particular piece does focus on mutations of the declarators which could fall victim to a certain amount of acorn restrictions?